### DISCUSSION TITLE
Introduce max-width for nodes

### DISCUSSION THREAD

**Discussion by tobi** [nikitabobko/AeroSpace#621] [feature-ideas]
Having windows go full-width on large monitors and ultrawides is jarring and undesired. There are a few tickets already on this, but I'd love to contribute my favorite way to solve this. And if the @nikitabobko is willing to merge this, I'm happy to offer a small bounty for the implementation, if that's desired. 

Introduce `default-node-max-width: 1280`. When set, this gives each new node a maximum width when it is created. On a screen that cannot be filled fully due to this constraint, a new window/node is centered with margins to the edges of the screen. I think would be an elegant improvement for the people that are bothered by this, and should introduce minimal amount of additional complexity to the system. Certainly avoids nasty late night, dark mode for hours, accidentally-flashbang-yourself-with-google.com on a 1800nits display problem which inspired this ticket 😅.

Some additional considerations:
- Floating windows are unaffected by this
- Nodes that are not connected to the root-container are unaffected by this
- We introduce a new layout command `layout tiling-fullwidth` that allows for hotkeys that disable the constraint. Example: `ctrl-alt-cmd-space = ['layout tiling-fullwidth tiling']`. Windows in this tiling-fullwidth state just ignore the maxwidth attribute
- The resize function should increase the node's maxwidth if its being resized beyond the current maxwidth's size so a node at 1200px width that gets a 'resize smart +200' sets its internal maxwidth to the new size. This means that moving the node later to another screen retains the session desired maxwidth
- resize should learn to set maxwidth for nodes specifically like `resize max-width 800`, this enables setting some types of windows to be constrained by default like so:
-
  ```
  [[on-window-detected]]
  if.app-id = 'com.googlecode.iterm2'
  run = ['layout tiling', 'resize max-width 800']
  ```
  
 I *think* that this is a horizontal only thing. Maybe there should be a vertical version of this as well, but I have a hard time coming up with a great use-case for this, so maybe best to just skip it.

**Comment by nikitabobko** [comment]
Thanks for sharing your vision!

> default-node-max-width

I'm curious if is it a mistake or is it intentional? The definition of "node" includes windows and containers. Restricting nodes max width will effectively restrict the root container width, which is no different from having big gaps.

My vision was different. I planned to introduce `windows-max-width`, and it would restrict only windows width, leaving containers width flexible. I lean towards restricting windows width because the more windows you have, the bigger becomes the total max width restriction (which should work nicely for the ultrawide monitor use case). Though I just realized that this vision has unfortunate feature interaction with mouse resizing.

---

Another thing is where to store "max width restriction". According to your description, you want to store it in the nodes/windows themselves. This way it's possible to resize nodes/windows independently

My vision was to have as many `windowsMaxWidth` as many monitors there are, and disallow resizing bigger than max width.

Since the primary use case is to improve experience for ultrawide monitors, it feels natural to me that `windowsMaxWidth` is a property of every monitor. And your idea to store "windows max width" in windows themselves isn't yet intuitive to me. 

I understand that "isn't yet intuitive to me" is a bad argument, and I will try add a comment later when I realize it. Maybe my intuition is afraid that every window will now have a property that is not directly observable (which feels too chaotic and random), or maybe it's something different, I didn't realize it yet.

> We introduce a new layout command `layout tiling-fullwidth` that allows for hotkeys that disable the constraint.

What would be the use case for disabling the constraint?

> resize should learn to set maxwidth for nodes specifically like `resize max-width 800`, this enables setting some types of windows to be constrained by default like so:

And I afraid that possibility to set `max-width` for every specific window encourages too much tinkering. I still yet lean more towards "windows max width per monitor" approach.

**Comment by adaviloper** [comment]
If a `window-max-width` is perhaps too restricting, have you considered perhaps providing a way to define layout "zones" in some way.

Something akin to defining proportioned areas that windows would be constrained to. Then you can have multiple proportions for the trees to work in.

For example, you could define multiple zones such as:

```
                               1/1 
+-------------------------------------------------------------+
|                                                             |
|                                                             |
|                                                             |
|                                                             |
|                                                             |
|                                                             |
|                                                             |
+-------------------------------------------------------------+

               1/1                            1/2
+-------------------------------------------------------------+
|                              |                              |
|                              |                              |
|                              |                              |
|                              |                              |
|                              |                              |
|                              |                              |
|                              |                              |
+-------------------------------------------------------------+

       1/4                     1/2                   1/4
+-------------------------------------------------------------+
|               |                              |              |
|               |                              |              |
|               |                              |              |
|               |                              |              |
|               |                              |              |
|               |                              |              |
|               |                              |              |
+-------------------------------------------------------------+

       1/3                     1/3                 1/3 
+-------------------------------------------------------------+
|                  |                     |                    |
|                  |                     |                    |
|                  |                     |                    |
|                  |                     |                    |
|                  |                     |                    |
|                  |                     |                    |
|                  |                     |                    |
+-------------------------------------------------------------+
```


etc etc

Could help to avoid having to manually calculate any different monitor sizes when switching between stations.

**Comment by tobi** [comment]
> Thanks for sharing your vision!
> 
> > default-node-max-width
> 
> I'm curious if is it a mistake or is it intentional? The definition of "node" includes windows and containers. Restricting nodes max width will effectively restrict the root container width, which is no different from having big gaps.

> My vision was different. I planned to introduce `windows-max-width`, and it would restrict only windows width, leaving containers width flexible. I lean towards restricting windows width because the more windows you have, the bigger becomes the total max width restriction (which should work nicely for the ultrawide monitor use case). Though I just realized that this vision has unfortunate feature interaction with mouse resizing.

I think the question is what we want to happen if the monitor width is larger than twice the max width. The first window will definitely be centered. What happens when the second window is added. Are both windows centered on the screen, or are both windows centered in their nodes, so effectively have a big gap between them? In effect this is the difference between css justify-content: centered or justify-content: space-between. I think strongly that centered is better. 

> Another thing is where to store "max width restriction". According to your description, you want to store it in the nodes/windows themselves. This way it's possible to resize nodes/windows independently
> 
> My vision was to have as many `windowsMaxWidth` as many monitors there are, and disallow resizing bigger than max width.
> 
> Since the primary use case is to improve experience for ultrawide monitors, it feels natural to me that `windowsMaxWidth` is a property of every monitor. And your idea to store "windows max width" in windows themselves isn't yet intuitive to me.

> 
> I understand that "isn't yet intuitive to me" is a bad argument, and I will try add a comment later when I realize it. Maybe my intuition is afraid that every window will now have a property that is not directly observable (which feels too chaotic and random), or maybe it's something different, I didn't realize it yet.

I think the case where the value will be overwritten will be pretty rare, so maybe your intuition is right. I'm thinking that web browsers should be max 1280px for most websites to look good, but sometimes you really want it to be wider using a specific website. This should just be possible without turning off the global setting. 



> 
> > We introduce a new layout command `layout tiling-fullwidth` that allows for hotkeys that disable the constraint.
> 
> What would be the use case for disabling the constraint?

Similar to the above: if you give a presentation in google slides, or are in a video chat that's best looked at full width screen. It would be bad if you have to turn off aerospace because you use a tool that needs the entire screen. 

> > resize should learn to set maxwidth for nodes specifically like `resize max-width 800`, this enables setting some types of windows to be constrained by default like so:
> 
> And I afraid that possibility to set `max-width` for every specific window encourages too much tinkering. I still yet lean more towards "windows max width per monitor" approach.

Totally losing the ability to make any window fullscreen feels challenging in day to day. VSCode flow state sessions, fullscreen presentations, fullscreen youtube video, IA Writer in zen mode, Video games in windowed mode. All those are things I wanted to temporarily use more than 1280px for a window recently. But maybe there is a simpler way to accomplish this?

**Comment by nikitabobko** [comment]
> In effect this is the difference between css justify-content: centered or justify-content: space-between. I think strongly that centered is better.

I agree

> if you give a presentation in google slides, or are in a video chat that's best looked at full width screen. It would be bad if you have to turn off aerospace because you use a tool that needs the entire screen.

I guess it makes sense... I never used an ultrawide monitor, so it's unclear to me how these apps look when they appear super wide. Do you mind attaching a screenshot?

**Comment by tobi** [comment]
don't have a good one as i'm on the road right now. But you can pretty much see the issue in the [review shots of this one](https://www.rtings.com/monitor/reviews/samsung/odyssey-neo-g9-g95na-s49ag95), which is the most common one. For what it's worth, i don't think that's an ultrawide only feature. While maybe not at laptop level- 30" screens would definitely benefit from a max-width cap.

**Comment by druvv** [comment]
> I guess it makes sense... I never used an ultrawide monitor, so it's unclear to me how these apps look when they appear super wide. Do you mind attaching a screenshot?

Here's a good example of Slack when it is super wide. As you can see, reading messages is pretty difficult because my eyes have to scan the entire width of the monitor. For a sense of scale, the horizontal width of my monitor is 32 inches (81 cm).

![Screenshot 2024-08-25 at 5 49 45 PM](https://github.com/user-attachments/assets/045dc5f7-aba8-4302-bf55-b0f397248233)

I've added pretty large outer gaps to compensate, but this isn't great because I lose screen real-estate when tiling multiple windows. Currently, aerospace doesn't completely solve my problem -- but with a max width it'll be perfect for me.

**Comment by bbrendon** [comment]
The problem is far worse on a 40 inch ultrawide. Also I see other great ideas https://github.com/nikitabobko/AeroSpace/issues/465 https://github.com/nikitabobko/AeroSpace/issues/60

**Comment by galligan** [comment]
Just wanted to +1 @tobi's request here, and stay in touch with this issue as it's currently the only thing holding me back from using AeroSpace, which I'm otherwise very excited about.

While some apps do a good job with keeping to reasonable line lengths of roughly 60-75 characters per line, some unfortunately don't make such accommodations—Slack, Spark, Mail.app, and Notes.app being notable examples. This makes using them on a screen like those described very difficult.

In my case, I've got a Pro Display XDR, and seeing windows like Finder, Notes, etc. full screen is jarring and offputting to the point where it's impractical to use them with AeroSpace. #552 would also help out in this instance.

Max width would be a great solution, as in a space where only one window is present, but max width is enforced, the surrounding space would simply be the desktop.

I'll also note that in some cases, perhaps restricted to certain spaces, max height would also be beneficial. That's because with a monitor as large as the Pro Display, keeping my vision centered in the middle is sometimes preferred, and a 100% height window can also be unwieldy.

Love the project so far @nikitabobko and looking forward to being able to use AeroSpace one day!

**Comment by nickcj931** [comment]
> If a `window-max-width` is perhaps too restricting, have you considered perhaps providing a way to define layout "zones" in some way.
> 
> Something akin to defining proportioned areas that windows would be constrained to. Then you can have multiple proportions for the trees to work in.
> 
> For example, you could define multiple zones such as:
> 

Having zones would be amazing as literally Aerosapce is pretty perfect for laptop size screen or non-wide screen displays, however its just shy of being useful for widescreens. And as others have mentioned creates a rather jarring experience. 

Being able to define zones means you can take control of large monitor estate and lock applications down based on those widescreen requirements, its sooo nearly there as i can transition my workspaces on docking my laptop to the new display automatically and keep some on the mac screen (which is just amazing!). 

Currently If i have e.g. 1 Browser session open on a workspace, when that moves to a widescreen it takes up the entire 5120x1440 estate which just makes it pointless, the only way to take control of this app window is to essentially open multiple other applications.  It would be nice to just have a "chill window" as per https://github.com/nikitabobko/AeroSpace/issues/60#issue-2044997998 where you have a centrally located browser window and nothing else around it to read some blogs etc. 

Also - great work @nikitabobko  I was about to throw my mac away as I was sick of dragging windows for a living in OSx!!

**Comment by nobodywasishere** [comment]
For me personally, I wish there was a way to define a default width an app opened in, instead of a max width. I tend to open and close apps (vs using workspaces) as it's less mental clutter for me, and every time I open Telegram or Messages, I need to resize them to their min width (as that's how small I like them when using them).

Something like this (which doesn't currently work):

```toml
[[on-window-detected]]
if.app-id = 'com.googlecode.iterm2'
run = 'resize width 300'
```

**Comment by webdz9r** [comment]
Tracking this like most - ultra wides are a first world problem, but a problem non-the-less... I'm going to share some lose concepts and ideas as I'm looking to use this.. 

I believe zones have been used to describe this concept? I could see myself setting up a workspace that I use for comms. I want Slack in zone 1, email in zone 2 and discord in zone 3 - this workspace would be ALT+C, additionally I'd like to setup these applications via configuration. 

This configuration feels like it could be defined to a monitor by default or specific to a workspace. I think someone [here](https://github.com/nikitabobko/AeroSpace/discussions/624) is trying to achieve the same result by defining workspace configurations. 

Maybe it feels like there is startup process which loads the app and then a bootstrap process that loads pre-configured workspaces accordingly. There are times I want to just have my normal floating windows without strict workspace bindings, so a workspace that is in floating mode with NO defined zones, basically a workspace that isn't managed by aerospace, free-form if you will

**Comment by dwhdai** [comment]
+1 on this. I came from Rectangle and really liked being able to center my browser to the middle third of my monitor when I was reading papers. With Aerospace I essentially have to fill the screen with more windows on the left/right which is undesirable because I want to just focus on the browser itself

**Comment by iminlikewithyou** [comment]
+1 !!! please! I don't use an ultrawide monitor often (mostly because of this problem), but when I use my Mac's internal display, I really appreciate being able to make my one-window setup a little smaller than filling the screen because I find that constraining the window a little is better for a lot of programs I use. Websites with video content similar to YouTube usually fit the content better this way, and it makes it a little easier to navigate! I use yabai with a custom script right now, and this is how my ideal tiling looks:

One window (constrained by a maximum width)
![CleanShot 2024-12-14 at 15 15 07@2x](https://github.com/user-attachments/assets/b348affd-8012-4a54-9b4e-875031ab2c46)
Multiple windows (no longer constrained)
![CleanShot 2024-12-14 at 15 16 25@2x](https://github.com/user-attachments/assets/7c20f38e-b28d-457a-8b02-f25d00f27289)

I totally and fully agree that it should be a property of the monitors, not workspaces, and definitely not windows (this is too much tinkering).

I just want to note that you guys are specifically noting "max-width", but I want to note that it SHOULD be "max pixels in the flow of direction". when using a horizontal flow, this is max-width. when using a vertical flow, this is max-height. this is notable for vertical monitors where a user of AeroSpace would like to keep the windows from stretching across the entire monitor, especially in the case of an ultrawide vertical monitor.

**Comment by FrenchGap** [comment]
+1 having a solution for this would remove the very last thorn I have with Aerospace usage today.
I was using rectangle prior to Aerospace and loved being able to place an app (usually my browser) to use 1/3 center of my ultrawide display when I wanted it to.

Here are some screenshots to give you a better idea, I use a 49 inch ultrawide display.

![2025-01-17_09-42_1](https://github.com/user-attachments/assets/22adbf08-0b90-47c4-bad7-dc292fe4dc87)
(One browser window open, occupying the whole ultrawide... it's unusable on most pages)

![2025-01-17_09-42](https://github.com/user-attachments/assets/12327975-c223-4e4b-9ce1-088888a1dfd1)
(With spacer windows on the sides, my current "fix" for this issue)

**Reply by epicwhale** [reply]
+1 this is the exact same problem I face - monitor is way too wide to have 1 browser window, so I want to max-width a browser window (or any window) on this monitor

**Comment by jlndk** [comment]
Posting my crude solution in case it might help someone.

I also have a 49" 32:9 ultrawide monitor and usually have two or three windows horizontally, where the current tiling algorithm works flawlessly.

However, in the case where I only have one window, I would prefer if it only took up the half the width (like a 16:9 monitor) and was centered horizontally.

I ended up writing a small bash script (triggered by `exec-and-forget` through a keybind) which floats the current window with the aerospace CLI and moves+resizes the window using applescript:


```bash
#!/usr/bin/env bash

# The output is like: "        UI Looks like: 1728 x 1117 @ 60hz"
DISPLAY_DETAILS=$(system_profiler SPDisplaysDataType | grep 'UI Looks like')

SCREEN_WIDTH=$(echo "$DISPLAY_DETAILS" | awk '{print $4}')
SCREEN_HEIGHT=$(echo "$DISPLAY_DETAILS" | awk '{print $6}')

if [ -z "$SCREEN_WIDTH" ] || [ -z "$SCREEN_HEIGHT" ]; then
    echo "Error: Could not determine screen dimensions."
    exit 1
fi

NEW_WIDTH=$((SCREEN_WIDTH / 2))
NEW_HEIGHT=$SCREEN_HEIGHT
NEW_X=$((SCREEN_WIDTH / 4)) # Position at 25% from the left to center it
NEW_Y=50 # Account for top bar

osascript -e "
try
    tell application \"System Events\" to tell window 1 of (process 1 where it is frontmost)
        try
      		set position to {$NEW_X, $NEW_Y}
            set size to {$NEW_WIDTH, $NEW_HEIGHT}
        end try
    end tell
on error errMsg
    display dialog \"AppleScript Error: \" & errMsg
end try
"
```

I originally tried to use `on-window-detected` with `exec-and-forget` to do this automatically based on the amount of windows in the workspace with this script:

```bash
#!/usr/bin/env bash

FOCUSED_WORKSPACE_INFO=$(aerospace list-workspaces --focused)

if [[ -z "$FOCUSED_WORKSPACE_INFO" ]]; then
    exit 0
fi

FOCUSED_WORKSPACE_ID=$(echo "$FOCUSED_WORKSPACE_INFO" | awk '{print $1}')

WINDOW_COUNT=$(aerospace list-windows --workspace "$FOCUSED_WORKSPACE_ID" | wc -l | awk '{print $1}')

# If there's only one window on the current workspace, center it,
# otherwise use the tiling layout
if [[ "$WINDOW_COUNT" -eq 1 ]]; then
    ./aerospace_center_window.sh # the script from above
else
    aerospace layout tiling
fi
```

but because of #20 I had to resort to a explicit keybind instead.
